The goal of this notebook is to compare two name-matching methods: `Standard Name Matching` and `Grouped Name Matching`.

The notebook is structured as follows:

- Preprocessing

- Summary Statistics: Overview of labeling outcomes across the dataset.

- Labeling Validation: Evaluating how well the methods categorize name pairs.

- Disagreement Analysis: Comparing the methods’ choices for best-match.

- Conclusion

- Future Improvements

- Limitations



# 1. Imports

In [1]:
#necessary libraries
from pathlib import Path
import pandas as pd  
import numpy as np  
import warnings  
from unidecode import unidecode
import re  
import matplotlib.pyplot as plt
import seaborn as sns
import time
from rapidfuzz import process, fuzz
import random
import scipy.stats as st
import math
from sklearn.metrics import classification_report, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns
 
#commands for better output readability 
pd.set_option('display.max_columns', None)  
pd.set_option('display.max_rows', None)  
warnings.filterwarnings("ignore", category=UserWarning, module='pandas')  
pd.set_option('display.max_colwidth', None)

In [2]:
#paths
project_dir=Path.cwd().parent.parent
processed_dir=project_dir/'data'/'processed'

matched_standard_file=processed_dir/'matched_standard.pkl'
matched_grouped_file=processed_dir/'matched_grouped.pkl'

df_standard=pd.read_pickle(matched_standard_file)  
df_grouped=pd.read_pickle(matched_grouped_file)  

# 2. Preprocessing

In [3]:
df_standard.head()

UK Sanction Programme  UK ID                      UK Name  \
0  ISIL (Da'esh) and Al-Qaida   6894   Abdul Rahman Mohamad Iqbal   
1                 Afghanistan   6895  Abdul Hai Hazem Abdul Qader   
2  ISIL (Da'esh) and Al-Qaida   6897          Abdul Man Am Saiyid   
3  ISIL (Da'esh) and Al-Qaida   6899    Tharwat Salah Shihata Ali   
4  ISIL (Da'esh) and Al-Qaida   6901        Majeed Abdul Chaudhry   

                 EU Name Match  Score  EU ID  String Length  Label  
0   Mohamad Iqbal Abdul Rahman    100   1004             26  match  
1  Abdul Hai Hazem Abdul Qader    100    505             27  match  
2          Abdul Man Am Saiyid    100    514             19  match  
3    Tharwat Salah Shihata Ali    100    796             25  match  
4        Majeed Chaudhry Abdul    100    641             21  match

In [4]:
df_grouped.head()

UK ID       UK Sanction Programme  \
0   6894  ISIL (Da'esh) and Al-Qaida   
1   6895                 Afghanistan   
2   6897  ISIL (Da'esh) and Al-Qaida   
3   6899  ISIL (Da'esh) and Al-Qaida   
4   6901  ISIL (Da'esh) and Al-Qaida   

                                                                                       UK Name  \
0  {Muqti, Abdurrahman, Mohamad, Jibril, Rahman, Fihiruddin, Fikiruddin, Abdul, A, Abu, Iqbal}   
1                                                                   {Hai, Abdul, Hazem, Qader}   
2                                               {Am, Saiyid, Al, Man, Abd, Manan, Abdul, Agha}   
3                           {Ali, Salah, Tharwat, Thirwat, Shihata, Abdallah, Shahata, Tarwat}   
4                                                             {Abdul, Chaudhry, Majeed, Majid}   

           UK Letters  \
0  {M, I, J, A, F, R}   
1           {A, H, Q}   
2           {M, A, S}   
3           {A, T, S}   
4           {M, C, A}   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Candidate EU IDs  \
0                                                                                                                                                                                                                                                                                                                          [630, 643, 1004, 3140, 4686, 5240, 5262, 5271, 5623, 6133, 6211, 6478, 6494, 6830, 6974, 7250, 113355, 115714, 117974, 123615, 125562, 126101, 127538, 129864, 130225, 133935, 134828, 135060, 136530, 136975, 138176, 145803, 146560, 147032, 150914, 159126, 162479, 162975, 165652, 166799, 167049, 167477, 171171, 172374, 172394]   
1                                                                                                                                                         [20, 505, 590, 591, 595, 603, 651, 661, 706, 709, 829, 842, 2193, 5270, 5416, 6078, 6093, 6095, 6113, 6130, 6207, 6224, 6240, 6312, 6583, 6616, 6694, 6873, 6887, 6898, 7085, 7146, 7368, 7386, 7447, 7501, 7513, 7586, 105305, 106138, 107163, 109900, 110138, 113244, 115145, 119228, 124306, 126530, 126538, 126554, 127417, 128186, 141403, 141872, 144861, 145430, 145486, 145521, 145691, 146428, 148293, 149178, 149390, 150680, 151879, 151903, 152704, 153011, 154125, 165849, 166346, 170630]   
2                                                                                           [54, 58, 76, 83, 103, 136, 143, 154, 156, 157, 176, 508, 514, 515, 516, 517, 522, 524, 526, 528, 545, 548, 553, 556, 581, 593, 595, 599, 600, 603, 604, 641, 643, 644, 656, 659, 661, 676, 696, 727, 733, 739, 758, 760, 765, 779, 781, 796, 826, 840, 931, 965, 1064, 1065, 1069, 1092, 1102, 1924, 2193, 2208, 2700, 3144, 3225, 3341, 3361, 3663, 3741, 3793, 3862, 4142, 5268, 5271, 5279, 5294, 5416, 5417, 5499, 5616, 5619, 5623, 5793, 5804, 6084, 6095, 6101, 6113, 6114, 6116, 6130, 6133, 6206, 6211, 6223, 6228, 6230, 6231, 6238, 6303, 6305, 6309, ...]   
3                                    [13, 20, 23, 54, 58, 67, 76, 83, 87, 98, 157, 507, 515, 516, 553, 604, 727, 733, 758, 765, 779, 786, 789, 796, 826, 829, 840, 1880, 1883, 1886, 1888, 1892, 1896, 1924, 2193, 2208, 2921, 3080, 3085, 3225, 3341, 3862, 5279, 5357, 5610, 5793, 6101, 6113, 6114, 6116, 6231, 6306, 6496, 6584, 6619, 6625, 6695, 6696, 6916, 6917, 6973, 6982, 7027, 7069, 7077, 7094, 7137, 7166, 7294, 7300, 7336, 7343, 7359, 736

# 3. Summary Statistics

## 3.1 Results

In [6]:
#standard matching label summary
counts_standard=df_standard['Label'].value_counts().sort_index()
proportions_standard=df_standard['Label'].value_counts(normalize=True).sort_index()
summary_standard=pd.DataFrame({'Count': counts_standard, 'Proportion': proportions_standard})

#grouped label summary
counts_grouped=df_grouped['Label'].value_counts().sort_index()
proportions_grouped=df_grouped['Label'].value_counts(normalize=True).sort_index()
summary_grouped=pd.DataFrame({'Count': counts_grouped, 'Proportion': proportions_grouped})

#print results
print("Standard Matching Label Summary:\n", summary_standard)
print("\nGrouped Matching Label Summary:\n", summary_grouped)

Standard Matching Label Summary:
                    Count  Proportion
Label                               
match               2699    0.748060
not match            271    0.075111
preliminary match    638    0.176829

Grouped Matching Label Summary:
                    Count  Proportion
Label                               
match               2390    0.662417
not match            438    0.121397
preliminary match    780    0.216186


## 3.2 Discussion

- The `standard method`is more decisive, assigning 74.8% of cases as matches and only 7.5% as non-matches, with 17.7% marked as preliminary matches. This suggests it tends to commit more often, possibly at the cost of precision in ambiguous cases.

- In contrast, the `grouped method` is more cautious, labeling only 66.2% as matches and increasing both non-matches (12.2%) and especially preliminary matches (21.6%), showing a preference for deferring decisions when confidence is lower.

# 4. Labeling Validation

## 4.1 Methodology

To assess the performance of each name-matching method in attributing labels, we manually evaluated a sample of name pairs. The goal was to determine how accurately each method classified matches, near matches, and non-matches. For each method, the UK IDs of incorrectly labeled pairs were recorded and used to generate confusion matrices and calculate precision metrics, enabling a direct comparison between the two approaches.

#### 4.1.1 *Sample*

To a accomplish this, a random sample of 250 UK IDs was selected and used for both matching methods to ensure a fair comparison. Although the ideal sample size for a ±5% margin of error at 95% confidence is 385, a smaller sample of 250 offered a practical balance between statistical rigor and manual labeling effort.

#### 4.1.2  *Evaluation Metrics*

As outlined in previous notebooks, we adopted a three-tier labeling strategy (Match, Not Match, and Preliminary Match)  designed to maximize confidence in automatic decisions while reserving uncertain cases for human review. This strategy was developed to reduce manual effort in the context of financial sanctions, where accuracy is critical and false positives carry significant consequences.

The most important evaluation metrics are:

- **Precision** in `match` and `not match`: If the system labels a pair as a match, it should truly be a match. Likewise, if it labels a pair as not match, it should be genuinely safe to ignore. These represent the most confident outcomes and must be highly reliable.

- **Recall** in `not match` (important but secondary): This metric indicates how many actual non-matches are confidently identified. Higher recall reduces the number of names falling into the Preliminary Match category, thereby lowering manual review workload. However, ensuring precision remains more critical.

We do not directly optimize for the `Preliminary Match` class and instead rely on it to catch edge cases and potential true matches that lack clear confidence.


We do not directly optimize for the `preliminary match` class. Instead, it serves as a catch-all for edge cases and uncertain true matches that fall below the confidence threshold. According to our strategy, it is perfectly acceptable for a real match to be classified as preliminary match, since these cases will still be manually reviewed and not missed.

#### 4.1.3 *Human Evaluation*

Determining whether two names correspond to the same person is a highly subjective task, complicated by different languages, language structures, spelling variations, and nuances that often require linguistic expertise to fully understand. To maintain consistency when visually evaluating matched names, the following guidelines were established:

- A pair is considered a match only if there are at least three common words; otherwise, it is classified as a preliminary match to reflect lower confidence.

- Since human evaluators may have varying familiarity with different languages, all words were weighted equally, with no distinction made between first names, middle names, or surnames. This approach was chosen to maintain consistency across unknown language structures.

- When only one word matches, the pair is usually classified as a non-match, except in cases where the matching word is a distinctive or unusually long name, in which case it may be considered a preliminary match.

## 4.2 Results

In [7]:
#create samples 
sample_size=250
sample_s=df_standard.sample(n=sample_size, random_state=5).sort_values(by='UK ID', ascending=True).reset_index(drop=True)
sample_g=df_grouped.sample(n=sample_size, random_state=5).sort_values(by='UK ID', ascending=True).reset_index(drop=True)

In [8]:
#inspect standard macthes
sample_s.head(250).sort_values(by='Score',ascending=False)

UK Sanction Programme  UK ID  \
0               ISIL (Da'esh) and Al-Qaida   6894   
1                              Afghanistan   6905   
2                              Afghanistan   6912   
3               ISIL (Da'esh) and Al-Qaida   6932   
4               ISIL (Da'esh) and Al-Qaida   7024   
5               ISIL (Da'esh) and Al-Qaida   7099   
6                              Afghanistan   7278   
7                              Afghanistan   7296   
8                              Afghanistan   7380   
9                              Afghanistan   7442   
10                             Afghanistan   7468   
11                             Afghanistan   7539   
12                             Afghanistan   7558   
13                                    Iraq   7580   
14                                    Iraq   7605   
15                                    Iraq   7611   
16                                    Iraq   8258   
17              ISIL (Da'esh) and Al-Qaida   8683   
18                                 Belarus   8904   
19              ISIL (Da'esh) and Al-Qaida   8983   
20              ISIL (Da'esh) and Al-Qaida   9225   
21                          Iran (Nuclear)  10634   
22              ISIL (Da'esh) and Al-Qaida  10667   
23              ISIL (Da'esh) and Al-Qaida  11097   
24                                    Iran  11778   
25                                    Iran  11785   
26                                    Iran  11795   
27                                    Iran  11799   
28                                   Syria  11933   
29                                   Syria  11934   
31                                   Syria  12216   
33                                   Syria  12418   
35                                   Syria  12507   
34                                   Syria  12469   
38                                   Syria  12642   
40                                   Syria  12705   
36              ISIL (Da'esh) and Al-Qaida  12631   
37                                   Syria  12635   
46   Democratic People's Republic of Korea  12868   
44              ISIL (Da'esh) and Al-Qaida  12808   
43                                   Syria  12776   
42                                   Syria  12769   
54                                  Russia  13017   
51                                  Russia  12953   
52                                  Russia  12962   
53                                   Syria  12997   
48                                  Russia  12923   
47                                  Russia  12891   
49                                  Russia  12950   
50                                  Russia  12951   
57                                  Russia  13068   
56                                  Russia  13041   
63              ISIL (Da'esh) and Al-Qaida  13196   
62                                 Somalia  13150   
61                             Afghanistan  13146   
60              ISIL (Da'esh) and Al-Qaida  13087   
59                                  Russia  13073   
58                                  Russia  13071   
121                                 Russia  14344   
115                                 Russia  14250   
116                                 Russia  14295   
99                                 Belarus  13962   
103                    Global Human Rights  14077   
101                                Belarus  13995   
64                                  Russia  13206   
65              ISIL (Da'esh) and Al-Qaida  13298   
66   Democratic People's Republic of Korea  13333   
68   Democratic People's Republic of Korea  13373   
69              ISIL (Da'esh) and Al-Qaida  13377   
70                Central African Republic  13379   
72                                  Russia  13394   
71                                   Syria  13381   
73                                   Syria  13413   
74                                   Syria  13497   
82                                 Myanmar  13

In [9]:
#store wrongly labeled UK IDs from the standard sample

change_to_match_s=[14535,16307,13838,14479,14918]	

change_to_not_match_s=[16072,12725,14807,15962,15440,12852,16068,15052,16014,15420,16309,15783,15651,16472,16782]		

change_to_preliminary_match_s=[7278,7558,8258,11778,11933,11934,12216,12418,12507,12469,12769,12769,13333,13413,13497,13505,16595,16542,15862,15626,16381,16367,14329]	

In [10]:
#inspect grouped macthes
sample_g[['UK ID', 'UK Name', 'Name Overlap','EU Name Match','Multi Score','Coverage Ratio','Raw Scores','Label']].sort_values(by='Multi Score', ascending=False).head(250)


UK ID  \
18    8904   
16    8258   
48   12923   
56   13041   
57   13068   
51   12953   
72   13394   
64   13206   
59   13073   
58   13071   
117  14305   
94   13927   
100  13974   
95   13939   
101  13995   
121  14344   
120  14333   
118  14322   
99   13962   
139  14540   
148  14691   
146  14686   
130  14412   
135  14513   
124  14365   
123  14364   
156  14851   
165  14971   
164  14968   
163  14949   
175  15178   
177  15236   
173  15171   
220  16217   
193  15633   
190  15571   
229  16381   
244  16723   
238  16602   
245  16731   
184  15387   
171  15166   
138  14538   
150  14703   
114  14222   
98   13957   
112  14209   
157  14865   
182  15318   
115  14250   
116  14295   
61   13146   
69   13377   
88   13841   
14    7605   
50   12951   
52   12962   
49   12950   
169  15085   
172  15170   
176  15234   
178  15246   
161  14925   
129  14406   
145  14674   
217  16169   
228  16373   
223  16316   
136  14529   
125  14388   
141  14582   
151  14732   
179  15257   
180  15277   
152  14739   
155  14850   
143  14634   
158  14884   
174  15175   
21   10634   
15    7611   
7     7296   
36   12631   
9     7442   
71   13381   
55   13039   
47   12891   
149  14700   
127  14395   
133  14473   
10    7468   
8     7380   
0     6894   
17    8683   
34   12469   
53   12997   
44   12808   
3     6932   
111  14176   
62   13150   
203  15851   
147  14690   
162  14937   
4     7024   
63   13196   
83   13769   
23   11097   
1     6905   
20    9225   
2     6912   
86   13814   
80   13625   
40   12705   
22   10667   
25   11785   
19    8983   
109  14149   
60   13087   
37   12635   
38   12642   
5     7099   
11    7539   
128  14397   
65   13298   
77   13592   
79   13618   
70   13379   
84   13770   
159  14905   
140  14558   
76   13539   
225  16327   
131  14421   
13    7580   
96   13945   
82   13692   
249  16818   
43   12776   
242  16710   
68   13373   
54   13017   
85   13809   
108  14123   
126  14390   
81   13662   
78   13598   
132  14426   
87   13838   
46   12868   
224  16320   
122  14345   
26   11795   
27   11799   
137  14535   
204  15854   
92   13894   
12    7558   
247  16772   
234  16542   
219  16208   
237  16595   
230  16444   
167  15052   
168  15069   
183  15342   
185  15419   
144  14663   
200  15818   
189  15526   
192  15626   
227  16367   
75   13505   
66   13333   
73   13413   
31   12216   
28   11933   
35   12507   
29   11934   
42   12769   
33   12418   
45   12852   
24   11778   
113  14215   
74   13497   
103  14077   
97   13946   
206  15862   
210  15988   
216  16165   
208  15935   
207  15920   
196  15717   
212  16026   
246  16758   
41   12725   
205  15860   
236  16593   
160  14918   
93   13895   
195  15651   
30   12045   
215  16072   
221  16307   
90   13860   
67   13347   
197  15783   
154  14807   
119  14329   
32   12230   
39   12650   
142  14602   
105  14091   
110  14160   
106  14098   
104  14088   
91   13870   
107  14101   
6     7278   
248  16782   
241  16700   
181  15312   
153  14799   
134  14479   
170  15139   
166  15006   
202  15838   
194  15638   
186  15420   
187  15440   
188  15445   
199  15802   
198  15794   
201  15823   
209  15962   
222  16309   
233  16490   
211  16014   
214  16068   
239  16659   
235  16589   
226  16357   
89   13859   
102  14011   
191  15593   
213  16067   
218  16195   
232  16485   
231  16472   
240  16698   
243  16715   

                                                                                                                                                                                                                                               UK Name  \
18                                                                                                                                                     {Vladimirovich, Uladzimiravich, Viktor, Sheiman, Sheyman, Ulazimiravich, Viktar, Uladimir

In [11]:
#store wrongly labeled UK IDs from the grouped sample

change_to_match_g=[]	

change_to_not_match_g=[15526,15651,14807]		

change_to_preliminary_match_g=[]	

In [12]:
#create a column with the true labels for each sample

sample_s['True Label']=sample_s['Label'].copy()
sample_g['True Label']=sample_g['Label'].copy()

sample_s.loc[sample_s['UK ID'].isin(change_to_match_s),'True Label']='match'
sample_s.loc[sample_s['UK ID'].isin(change_to_not_match_s),'True Label']='not match'
sample_s.loc[sample_s['UK ID'].isin(change_to_preliminary_match_s),'True Label']='preliminary match'

sample_g.loc[sample_s['UK ID'].isin(change_to_match_g),'True Label']='match'
sample_g.loc[sample_s['UK ID'].isin(change_to_not_match_g),'True Label']='not match'
sample_g.loc[sample_s['UK ID'].isin(change_to_preliminary_match_g),'True Label']='preliminary match'

In [13]:
class_labels=['not match','preliminary match', 'match']


#calculate classification report
report_s=classification_report(
    sample_s['True Label'], 
    sample_s['Label'], 
    labels=class_labels, 
    output_dict=True
)

#compute confusion matrix
confusion_matrix_s=confusion_matrix(sample_s['True Label'],sample_s['Label'], labels=class_labels)

#extract relevant metrics
precision_match_s=report_s['match']['precision']
precision_not_match_s=report_s['not match']['precision']
recall_not_match_s=report_s['not match']['recall']

#print results
print("Confusion Matrix:")
print(confusion_matrix_s)

print("\nClassification Report Summary:")
print(f"'Match' Precision:     {precision_match_s}")
print(f"'Not Match' Precision: {precision_not_match_s}")
print(f"'Not Match' Recall:    {recall_not_match_s}")

Confusion Matrix:
[[ 21  14   1]
 [  0  38  21]
 [  0   5 150]]

Classification Report Summary:
'Match' Precision:     0.872093023255814
'Not Match' Precision: 1.0
'Not Match' Recall:    0.5833333333333334


In [14]:
class_labels=['not match','preliminary match', 'match']


#calculate classification report
report_g=classification_report(
    sample_g['True Label'], 
    sample_g['Label'], 
    labels=class_labels, 
    output_dict=True
)

#compute confusion matrix
confusion_matrix_g=confusion_matrix(sample_g['True Label'],sample_g['Label'], labels=class_labels)

#extract relevant metrics
precision_match_g=report_g['match']['precision']
precision_not_match_g=report_g['not match']['precision']
recall_not_match_g=report_g['not match']['recall']

#print results
print("Confusion Matrix:")
print(confusion_matrix_g)

print("\nClassification Report Summary:")
print(f"'Match' Precision:     {precision_match_g}")
print(f"'Not Match' Precision: {precision_not_match_g}")
print(f"'Not Match' Recall:    {recall_not_match_g}")

Confusion Matrix:
[[ 39   3   0]
 [  0  52   0]
 [  0   0 156]]

Classification Report Summary:
'Match' Precision:     1.0
'Not Match' Precision: 1.0
'Not Match' Recall:    0.9285714285714286


## 4.3 Discussion

- The `grouped method` achieved perfect precision for both match and not match classes, and high not match recall (0.91), confirming it only commits to clear cases, consistent with its more conservative behavior observed earlier.

- The `standard method` showed slightly lower precision for match (0.87) and weak recall for not match (0.58), meaning it’s more willing to classify aggressively but at the cost of mislabeling.

This aligns with prior observations: the `standard method` is more decisive but riskier, while the `grouped method` is cautious and favors accuracy.

# 5. Disagreement Analysis 

In [15]:
standard_map=df_standard[['UK ID', 'UK Name', 'EU ID', 'EU Name Match', 'Label']].rename(
    columns={'EU ID': 'Standard Method', 'Label': 'Standard Label', 'EU Name Match': 'EU Name (Standard)','UK Name':'UK Name (Standard)'}
)

grouped_map=df_grouped[['UK ID','UK Name', 'EU ID', 'Name Overlap','EU Name Match', 'Label']].rename(
    columns={'EU ID': 'Grouped Method', 'Label': 'Grouped Label', 'EU Name Match': 'EU Name (Grouped)','UK Name':'UK Name (Grouped)'}
)

df_comparison=pd.merge(standard_map, grouped_map, on='UK ID', how='inner')

#ensure consistent types
df_comparison['Standard Method']=df_comparison['Standard Method'].astype('Int64')
df_comparison['Grouped Method']=df_comparison['Grouped Method'].astype('Int64')

def compare_ids(row):
    val1=row['Standard Method']
    val2=row['Grouped Method']
    if pd.isna(val1) or pd.isna(val2):
        return False
    return val1==val2

df_comparison['Same EU ID'] = df_comparison.apply(compare_ids, axis=1)

#filter out cases where the names were matched to the same EU ID + cases were a match was not found
df_mismatch=df_comparison[df_comparison['Same EU ID']==False]
df_mismatch=df_mismatch[~((df_mismatch['Standard Label']=='not match') & 
                            (df_mismatch['Grouped Label']=='not match'))]

#reorder columns for clarity
df_mismatch=df_mismatch[[
    'UK ID', 'UK Name (Standard)',
    'Standard Method', 'EU Name (Standard)', 'Standard Label',
    'UK Name (Grouped)','Grouped Method', 'Name Overlap','EU Name (Grouped)', 'Grouped Label'
]]

#print results
print(f"Total meaningful mismatches in EU IDs: {len(df_mismatch)}")
df_mismatch.head(100)

Total meaningful mismatches in EU IDs: 588


UK ID                   UK Name (Standard)  Standard Method  \
37     6994               Khalid Sheikh Mohammed              191   
43     7016                         Abu Mohammed             4141   
105    7354                        Mohammad Naim              654   
196    7860                          Ali Barkani             6827   
207    7886                         Usama Hamdan           152920   
209    7888                             Abu Umar             7482   
218    8248  Ali Barzan Ibrahim Hasan Al Tikriti             1898   
254    8709                      Jules Mutebutsi             3189   
262    8720                            Abu Akram             5618   
300   10637                      Ali Hoseynitash             5270   
303   10643                  Ali Mohammad Naderi             7091   
307   10647                         Ahmad Vahidi             5280   
313   10666                 Redendo Cain Dellosa             5261   
326   10834                            Abdul Haq              661   
335   10929                   Yassin Mohamed Ali             7091   
342   11030                        Chol Hae Hyon           115172   
365   11235                     Mohammad Pakpour           127411   
377   11635           Aisha Muhammed Abdul Salam             7547   
385   11644                        Khaled Tohami           118998   
386   11646       Sayyid Mohammed Qadhafi Al Dam           109906   
387   11647             Mohammed Muammar Qadhafi             6111   
399   11703                  Abdelmajid Al Gaoud             1102   
402   11708                     Abdallah Mansour             3960   
426   11800               Seyed Mohammad Soltani             5419   
433   11901                          Ali Mamlouk             6304   
442   11912                       Amjad Al Abbas             6309   
446   11933                      Mohammed Hamcho             7543   
451   12013                    Zoulhima Chaliche             6477   
453   12015                       Khaled Qaddour             7544   
455   12020                  Abd Al Rahim Bashir           105748   
467   12045             Mohammed Saeed Bukhaytan           107009   
474   12053                         Samir Hassan             6509   
486   12176                      Hassan Akharian             6583   
506   12199              Mohammad Bagher Bagheri             6606   
517   12211                  Ali Abdullah Ayyoub             6623   
519   12213                           Aous Aslan             6625   
525   12221                         Nazih Zareer           152815   
531   12230                     Seyed Shamseddin              594   
536   12240                   Mohammad Mohammadi             6656   
542   12415                    Ibrahim Al Hassan             6825   
571   12481                        Nasser Al Ali             6891   
583   12510                         Faysal Abbas             5619   
606   12650                    Ali Asghar Hejazi             6972   
607   12651                           Ali Saeedi             6973   
641   12725                          Ahmed Salem             1065   
643   12729                         Issam Hallaq             7092   
644   12730                      Ezzedine Ismail             7093   
649   12765             Mohammed Abdul Al Sayyed           109906   
659   12781                Adnan Abdou Al Sikhny             7143   
684   12852                    Mohammad Sarafraz              653   
698   12872                           Abu Hamawi              926   
704   12894             Viktor Pavlovych Pshonka             7270   
752   12971              Igor Nikolaevich Bezler           138295   
767   12995                            Malik Ali             7367   
809   13084                             Abu Anas              537   
811   13086      Abu Mohammad Al Adnani Al Shami           139991   
813   13088             Hamid Hamad Hamid Al Ali             7447   
907 

In [16]:
standard_wins=[8248,8709,11901,11912,12176,12199,12211,12213,12729,12730,12781,13088] #12
grouped_wins=[7354,7860,7888,8720,11635,11647,11800,12221,12651,12725,12765,12872,2971,13084,13301,13369] #16
draw=[6994,7016,7886,10637,10643,10647,10834,10929,11030,11235,11644,11646,11708,11933,12013,12015,12020,12045,12053,12230,12240,12415,12481,
     12510,12650,12852,12894,12995,13086,13233,13252,13347] #32

It is worth further inspecting the cases in which the standard approach clearly picked a better match than the grouped approach for the UK ID to try to comprehend the limitations of the algorithm created to match the names. For thsi, we printed thsoe UK IDs fro mismatch and also the df_grouped for those UK IDs to see if there is some relationship between the metrcis and those errors. 

In [17]:
df_mismatch[df_mismatch['UK ID'].isin(standard_wins)]

UK ID                   UK Name (Standard)  Standard Method  \
218   8248  Ali Barzan Ibrahim Hasan Al Tikriti             1898   
254   8709                      Jules Mutebutsi             3189   
433  11901                          Ali Mamlouk             6304   
442  11912                       Amjad Al Abbas             6309   
486  12176                      Hassan Akharian             6583   
506  12199              Mohammad Bagher Bagheri             6606   
517  12211                  Ali Abdullah Ayyoub             6623   
519  12213                           Aous Aslan             6625   
643  12729                         Issam Hallaq             7092   
644  12730                      Ezzedine Ismail             7093   
659  12781                Adnan Abdou Al Sikhny             7143   
813  13088             Hamid Hamad Hamid Al Ali             7447   

                      EU Name (Standard)     Standard Label  \
218  Ali Barzan Ibrahim Hasan Al Tikriti              match   
254                      Jules Mutebutsi              match   
433                          Ali Mamlouk              match   
442                       Amjad Al Abbas              match   
486                      Hassan Akharian              match   
506              Mohammad Bagher Bagheri              match   
517                  Ali Abdullah Ayyoub              match   
519                       Aous Ali Aslan  preliminary match   
643                         Issam Hallaq              match   
644                      Ezzedine Ismail              match   
659                Adnan Abdou Al Sikhny              match   
813             Hamid Hamad Hamid Al Ali              match   

                                                    UK Name (Grouped)  \
218                        {Hasan, Tikriti, Ibrahim, Barzan, Ali, Al}   
254                            {Mutebuzi, Jules, Mutebusi, Mutebutsi}   
433                                            {Ali, Mamlouk, Mamluk}   
442                                                {Amjad, Abbas, Al}   
486                                       {Hossein, Hassan, Akharian}   
506                                       {Mohammad, Bagher, Bagheri}   
517       {Ayoub, Ayob, Ayyub, Abdullah, Ayyoub, Abdallah, Ayub, Ali}   
519                                                     {Aslan, Aous}   
643                     {Hallak, Issam, Halaq, Barber, Hallaq, Essam}   
644                                 {Ismail, Ezzedine, Hamra, Ismael}   
659  {Sukhny, Sekhni, Abdou, Sekhny, Sikhny, Adnan, Abdo, Sikhni, Al}   
813                                           {Ali, Hamid, Hamad, Al}   

     Grouped Method                            Name Overlap  \
218             101  [Tikriti, Al, Ibrahim, Barzan, Hassan]   
254          168504                                 [Jules]   
433             508                                  [Alfi]   
442            7359                             [Al, Abbas]   
486            1892                       [Hassan, Hussein]   
506            3808                      [Mohammad, Bagher]   
517             154                              [Abdallah]   
519          145712                               [Absalan]   
643          146470                                   [Isa]   
644            6236                               [Esmaili]   
659              13                              [Al, Abou]   
813            7262                      [Al, Hamid, Hamza]   

                                                                                       EU Name (Grouped)  \
218                                                               {Hassan, Tikriti, Ibrahim, Barzan, Al}   
254                                    {Mbitshemunda, Habyarimana, Jean, Mulumba, Jules, Claude, Mbitse}   
433                                  {Masri, Saleh, Ahmed, Mohamed, Abdullah, Alfi, Mariam, El, Abu, Al}   
442                                                                                {Suleiman, Abba

In [18]:
df_grouped_errors=df_grouped[df_grouped['UK ID'].isin(standard_wins)]
df_grouped_errors=df_grouped_errors.sort_values('UK ID')
df_grouped_errors.head(30)

UK ID             UK Sanction Programme  \
218   8248                              Iraq   
254   8709  Democratic Republic of the Congo   
433  11901                             Syria   
442  11912                             Syria   
486  12176                              Iran   
506  12199                              Iran   
517  12211                             Syria   
519  12213                             Syria   
643  12729                             Syria   
644  12730                             Syria   
659  12781                             Syria   
813  13088        ISIL (Da'esh) and Al-Qaida   

                                                              UK Name  \
218                        {Hasan, Tikriti, Ibrahim, Barzan, Ali, Al}   
254                            {Mutebuzi, Jules, Mutebusi, Mutebutsi}   
433                                            {Ali, Mamlouk, Mamluk}   
442                                                {Amjad, Abbas, Al}   
486                                       {Hossein, Hassan, Akharian}   
506                                       {Mohammad, Bagher, Bagheri}   
517       {Ayoub, Ayob, Ayyub, Abdullah, Ayyoub, Abdallah, Ayub, Ali}   
519                                                     {Aslan, Aous}   
643                     {Hallak, Issam, Halaq, Barber, Hallaq, Essam}   
644                                 {Ismail, Ezzedine, Hamra, Ismael}   
659  {Sukhny, Sekhni, Abdou, Sekhny, Sikhny, Adnan, Abdo, Sikhni, Al}   
813                                           {Ali, Hamid, Hamad, Al}   

          UK Letters  \
218  {H, T, I, B, A}   
254           {M, J}   
433           {M, A}   
442              {A}   
486           {H, A}   
506           {M, B}   
517              {A}   
519              {A}   
643     {H, I, E, B}   
644        {H, I, E}   
659           {S, A}   
813           {H, A}   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Candidate EU IDs  \
218                                                          [13, 25, 33, 39, 54, 65, 69, 87, 89, 98, 99, 101, 506, 544, 706, 829, 943, 945, 1016, 1888, 1892, 1894, 1896, 1898, 1900, 1920, 1922, 2325, 3080, 3081, 3082, 3084, 3085, 3225, 3340, 4140, 6047, 6049, 6060, 6063, 6074, 6078, 6079, 6080, 6082, 6177, 6457, 6480, 6612, 6825, 6889, 7150, 7294, 7370, 7477, 7478, 7498, 7601, 7602, 105748, 106534, 106538, 107679, 107693, 109872, 110845, 111553, 112075, 115225, 118547, 121825, 125574, 126488, 126714, 127732, 130197, 131720, 133626, 134079, 134380, 135748, 135879, 136340, 138684, 142836, 142906, 146768, 149736, 152809, 152900, 154528, 155925, 156392, 161670, 162355, 166170, 166475, 169200, 172452, 172959]   
254                                                                                                                                                                                                               [614, 623, 630, 649, 690, 851, 1923, 4141, 5271, 6108, 6211, 6508, 6592, 6593, 6597, 6648, 6656, 6822, 6830, 6871, 7141, 7201, 7206, 7495, 7500, 105596, 106457, 107816, 108458, 109183, 109876, 110276, 115228, 115965, 117521, 117533, 123624, 124315, 124573, 124987, 125486, 127549, 131237, 134007, 134115, 134292, 134388, 134832, 135640, 138155, 138489, 138504, 138841, 140489, 145794, 146410, 148363, 149393, 149904, 149914, 153611, 155933, 156306, 159833, 161870, 164367

## 5.1 Discussion

Out of the 3,608 unique UK individuals, the standard and grouped methods matched to different EU IDs in 587 cases. This already excludes instances where both methods selected different matches but classified them as not match.

From a manual review of the first 60 disagreement cases:

- 32/60 involved different but equally reasonable matches.

- 16/60 were cases where the grouped method found a significantly better match than the standard method — as expected.

- 12/60 showed the standard method produced a clearly better match, which was surprising given its simplicity as a baseline method.

Upon further investigation, those 12 unexpected failures revealed several limitations and blind spots in the grouped algorithm, including:

1. **Missing matches in candidate pool**: In some cases, the correct EU ID selected by the standard method wasn’t even included in the grouped method’s candidate list. This likely happened because the similarity threshold in the grouped method was too strict, causing it to overlook valid matches that scored just below the cutoff.
    
    - Example (UK ID 12729): `Issam Hallaq` was correctly matched by the standard method to EU ID 7092 (`Issam Hallaq`). However, the grouped method matched it to “Isa Bayat`, a completely different person. EU ID 7092 wasn’t even part of the candidate list, indicating the threshold filtered it out.
<br>

2. **Confusion within same family**: The grouped method sometimes confused individuals from the same family or name group. This was often due to short first names being ignored or underweighted—yet those are sometimes the only features that differentiate otherwise similar names.

   - Example (UK ID 8248): `Ali Barzan Ibrahim Hasan Al Tikriti` was correctly matched by the standard method. The grouped method picked another `Al Tikriti` family member, likely because it did not give enough importance to the short but distinctive token `Ali`.
<br>

3. **Over-simplification of similar words**: At times, the grouped method treated distinct names as if they were the same, especially when spellings were similar. This led to mismatches between people whose names look alike but are not actually the same.

   - Example (UK ID 12199): `Mohammad Bagher Bagheri` was matched correctly by the standard method. The grouped method, however, selected `Mohammad Bagher` instead, likely treating `Bagher` and `Bagheri` as mere variants of the same name, rather than distinct individuals.
<br>

4. **Too much weight on common words**: The algorithm occasionally gave excessive weight to generic name elements such as “`Mohammad`, `Al`, `Abu` etc. These frequent terms added noise to the comparison and sometimes led the model to overlook more distinctive name components.



# 6. Conclusion

The grouped method demonstrates clear superiority in accurately labeling matches, prioritizing precision and minimizing false positives. This cautious approach inevitably increases the number of cases requiring human review, which is expected given the inherent complexity of name matching across diverse languages and variations.

When it comes to selecting the best match, the two methods are broadly comparable. They disagreed in about 16% of cases, but roughly half of those disagreements involved equally valid alternatives. In the remaining cases, each method outperformed the other on some occasions, reflecting different strengths and weaknesses.

Notably, the few instances where the standard method outperformed the grouped method revealed blind spots in the grouped algorithm—limitations that stem from its design and thresholding, which sometimes cause it to miss the best matches.

Overall, while the grouped method excels in label accuracy and reducing false matches by capturing nuanced name information, fundamental algorithmic improvements are essential to ensure consistent performance. Without these, the method performs well most of the time but can fail drastically in certain cases.

# 7. Future Improvements

- Explore lowering the similarity threshold used in pre-selecting EU candidates to reduce the chance of missing good matches, while keeping computation time reasonable.

- Incorporating linguistic heuristics, especially for Arabic and Russian names, which are common in sanctions datasets. For example:

    - Assigning lower weight to very common names like `Muhamed` and its variants to minimize noise.

    - Accounting for patronymic name structures in Russian names, such as `Alekseyveech`being derived from the given name `Aleksey` (meaning 'son of').


# 8. Key Limitations

1. **Subjectivity of the task:** Despite efforts to maintain consistency and minimize bias, evaluating name matches remains inherently subjective due to linguistic nuances and variations. However, in real-world sanctions screening, name matching is only one part of a broader verification process that includes multiple personal details, which helps mitigate this subjectivity.

3. **Sample Size Constraints:** The full dataset included 3,608 individuals, but manual evaluations were limited to smaller samples of 250 and 60 pairs due to resource constraints. These small sample sizes reduce the statistical robustness of the validation and disagreement analyses.

5. **Dataset Choice and Risk of Overfitting:** To test matching methods effectively, two similar but not identical sanctions lists (UK and EU) were chosen, resulting in substantial overlap of individuals. While this allowed for a meaningful matching task, it may introduce overfitting risks, limiting the generalizability of results to other datasets or contexts.

# 9. Output

Since a validated sample of the grouped matching results has already been created, we can now use it for further analysis. The final datasets were saved as follows:

In [19]:
sample_g.head()

UK ID       UK Sanction Programme  \
0   6894  ISIL (Da'esh) and Al-Qaida   
1   6905                 Afghanistan   
2   6912                 Afghanistan   
3   6932  ISIL (Da'esh) and Al-Qaida   
4   7024  ISIL (Da'esh) and Al-Qaida   

                                                                                                     UK Name  \
0                {Fihiruddin, Abdul, Jibril, Fikiruddin, Iqbal, Rahman, Abdurrahman, Mohamad, Abu, Muqti, A}   
1                                                          {Abdul, Zakir, Akhund, Mullah, Sahib, Bari, Haji}   
2                                                                               {Aliza, Rauf, Abdul, Khadem}   
3  {Mohammed, Ismail, Al, Uthman, Filistini, Takfiri, Omar, Qatada, Mahmoud, Samman, Umar, Abu, Umr, Othman}   
4                                  {Aouadi, Belgacem, Belkacem, Mohamed, Hannachi, Fathi, Abdallah, Ben, Al}   

                    UK Letters  \
0           {M, J, F, I, A, R}   
1           {M, S, H, B, A, Z}   
2                    {K, A, R}   
3  {M, Q, U, S, F, O, T, I, A}   
4              {M, H, F, B, A}   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Candidate EU IDs  \
0                                                                                                                                                                                                                                                                                                                                                                                              [630, 643, 1004, 3140, 4686, 5240, 5262, 5271, 5623, 6133, 6211, 6478, 6494, 6830, 6974, 7250, 113355, 115714, 117974, 123615, 125562, 126101, 127538, 129864, 130225, 133935, 134828, 135060, 136530, 136975, 138176, 145803, 146560, 147032, 150914, 159126, 162479, 162975, 165652, 166799, 167049, 167477, 171171, 172374, 172394]   
1                                                                                                                 [25, 46, 54, 103, 514, 515, 528, 553, 556, 593, 595, 599, 600, 603, 604, 661, 727, 750, 927, 928, 937, 965, 966, 1016, 1065, 1069, 1085, 1102, 1894, 1896, 2193, 2208, 3084, 3144, 3225, 3341, 3361, 3663, 3808, 4142, 5294, 5295, 5416, 5417, 5529, 5616, 5619, 5625, 6044, 6047, 6049, 6052, 6053, 6054, 6073, 6074, 6078, 6079, 6084, 6085, 6090, 6116, 6130, 6213, 6223, 6238, 6309, 6484, 6494, 6607, 6608, 6616, 6695, 6898, 6947, 6982, 7065, 7069, 7074, 7166, 7199, 7223, 7231, 7443, 7472, 7474, 7475, 7477, 7483, 7484, 7496, 7597, 106138, 106144, 106385, 106389, 106534, 106538, 106584, 106654, ...]   
2  [143, 146, 513, 659, 676, 719, 724, 786, 849, 1096, 2921, 3062, 3140, 3342, 3793, 5266, 5925, 6212, 6215, 6216, 6240, 6582, 6584, 6650, 6945, 6974, 6989, 6993, 7085, 7221, 7229, 7237, 7250, 7291, 7361, 7368, 7405, 7435, 7455, 7456, 7504, 7513, 7519, 7523, 7567, 107818, 108640, 109244, 109980, 110174, 110425, 113326, 117485, 117974, 119548, 119656, 123615, 125538, 125637, 126338, 126507, 126698, 127526, 127528, 127726, 128270, 128279, 129815, 129836, 129884, 130225, 130355, 132522, 133907, 133959, 134035, 134083, 134107, 134171, 134175, 134340, 134484, 134488, 134532, 134556, 134684, 134860, 134872, 134876, 134880, 134884, 134892, 134900, 134920, 134924, 134928, 134932, 134936, 134964, 134976, ...]   
3                               

In [21]:
df_grouped.to_pickle(processed_dir/'matched_grouped_final.pkl')
sample_g.to_pickle(processed_dir/'matched_grouped_labeled_sample.pkl')